### PR TITLE
Move shared cache pre-allocation and support macOS

### PR DESCRIPTION
**Pull request by jasontedor**
This commit moves the shared cache pre-allocation code out of bootstrap, and instead to the searchable snapshots code. We got out of our way to only grant permissions to a specific library used for the pre-allocation.

Additionally, to ensure our interfaces are sound, we add a macOS implementation based on fcntl and ftruncate.

Relates #68858

### CODE REVIEW

**Comment by jasontedor** [comment]
The main purpose of this relocation is to keep this plugin-specific code out of bootstrap. Additionally, for some complicated reasons, we can't touch the roles until *after* plugins are done being loaded. This means that we can't make the validation fo these settings dependent on the node roles unless we relocate these settings out of bootstrap so that they are not touched until after plugins are done being loaded. This gives us a second reason to move this code, so we take those steps now.

**Review by rjernst — APPROVED** [review]
The plugin+security policy side LGTM.

**Review by ywelsch — COMMENTED** [review]
I've done a first pass, and pushed a commit to fix precommit (as best I could).

I'm wondering if there's some test we can add to make sure this actually works on MacOS / Linux besides manual testing.

File: x-pack/plugin/searchable-snapshots/preallocate/src/main/java/org/elasticsearch/xpack/searchablesnapshots/preallocate/MacOsPreallocator.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public boolean available() {
+        return Natives.NATIVES_AVAILABLE;
+    }
+
+    @Override
+    public int preallocate(final int fd, final long currentSize /* unused */ , final long fileSize) {
+        // the Structure.ByReference constructor requires access to declared members
+        final Natives.Fcntl.FStore fst = AccessController.doPrivileged((PrivilegedAction<Natives.Fcntl.FStore>) Natives.Fcntl.FStore::new);
+        fst.fst_flags = Natives.Fcntl.F_ALLOCATEALL;
```
[the remark this answers is not in the document]
Comment (jasontedor) [reply x-pack/plugin/searchable-snapshots/preallocate/src/main/java/org/elasticsearch/xpack/searchablesnapshots/preallocate/MacOsPreallocator.java]: Thanks for the suggestion; I pushed fae8f993483f923b2841e2712f26eb824a0b9e36.

**Comment by jasontedor** [comment]
@ywelsch Thanks for making a pass, and especially for addressing precommit. 🙏

I took your suggestion to first attempt to allocate contiguously on macOS.

I agree on adding testing, I'll do that in a follow-up.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**